### PR TITLE
JAMES-2273 James should warn when the jwt token is invalid or empty

### PR DESCRIPTION
**Pull request by quynhn**

### CODE REVIEW

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPConfiguration.java [hunk 1]
```diff
@@ -84,6 +86,11 @@ public JMAPConfiguration build() {
             Preconditions.checkState(!enabled.get() || !Strings.isNullOrEmpty(secret), "'secret' is mandatory");
             return new JMAPConfiguration(enabled.get(), keystore, secret, jwtPublicKeyPem, port);
         }
+
+        private boolean validPublicKey(Optional<String> jwtPublicKeyPem) {
+            PublicKeyReader reader = new PublicKeyReader();
+            return !jwtPublicKeyPem.isPresent() || reader.fromPEM(jwtPublicKeyPem).isPresent();
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPConfiguration.java]: You have two checks here:
 - Does the JWT key is valid? -> this one belongs in JWTConfiguration.java
 - JMAP requires a JWT key -> this check should be performed in Jmap configuration

**Comment by quynhn** [comment]
Sorry for push force because there are unchanged files after all

File: server/protocols/jwt/src/main/java/org/apache/james/jwt/JwtConfiguration.java [hunk 2]
```diff
@@ -21,13 +21,21 @@
...
     private final Optional<String> jwtPublicKeyPem;
 
     public JwtConfiguration(Optional<String> jwtPublicKeyPem) {
+        Preconditions.checkState(validPublicKey(jwtPublicKeyPem), "The provided public key is not valid");
         this.jwtPublicKeyPem = jwtPublicKeyPem;
     }
 
+    private boolean validPublicKey(Optional<String> jwtPublicKeyPem) {
+        PublicKeyReader reader = new PublicKeyReader();
+        return !jwtPublicKeyPem.isPresent() || reader.fromPEM(jwtPublicKeyPem).isPresent();
```
Comment (chibenwa) [inline server/protocols/jwt/src/main/java/org/apache/james/jwt/JwtConfiguration.java]: I find this boolean close hard to read....

Maybe a more functional style might be easier?

I propose:

```
PublicKeyReader reader = new PublicKeyReader();
jwtPublicKeyPem.map(value -> reader.fromPEM(Optional.of(value)).isPresent())
    .orElse(DEFAULT_VALID);
```

File: server/protocols/jwt/src/test/java/org/apache/james/jwt/JwtConfigurationTest.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    @Test
+    public void getJwtPublicKeyPemShouldReturnEmptyWhenEmptyPublicKey() throws Exception {
+        JwtConfiguration jwtConfiguration = new JwtConfiguration(Optional.empty());
+
+        assertThat(jwtConfiguration.getJwtPublicKeyPem()).isNotPresent();
+    }
+
+    @Test
+    public void constructorShouldThrowWhenNullPublicKey() throws Exception {
+        assertThatThrownBy(() -> new JwtConfiguration(Optional.of(null)))
```
Comment (chibenwa) [inline server/protocols/jwt/src/test/java/org/apache/james/jwt/JwtConfigurationTest.java]: It's Optional.of(null) which is throwning, not JwtConfiguration constructor. This test can be transformed to:

```
assertThatThrownBy(() -> new JwtConfiguration(null))
    .isInstanceOf(NullPointerException.class);
```

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
```
[5e6eb3c66cb599c75b0a0ae404828d645452832d] [ERROR]   Java heap space
```

test this please

**Comment by quynhn** [comment]
```
[ERROR] Errors: 
[ERROR]   A scoping block is already in progress
[ERROR]   Java heap space
[ERROR]   A scoping block is already in progress
[ERROR]   A scoping block is already in progress
[ERROR]   A scoping block is already in progress
[ERROR]   A scoping block is already in progress
[ERROR]   A scoping block is already in progress
```

**Comment by quynhn** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**